### PR TITLE
Parse with file name rather than file contents

### PR DESCRIPTION
**Pull request by LuigiPiucco**
I modified this repo for my own use-case, but it's a very small change and I
think it's a generally useful one. Specifically, I'm trying to wrap DOOM Emacs
in Nix, by replacing straight.el. DOOM configs have 3 files, and each has its
meaning, it's not just package definitions, and package related pieces may come
in a few different forms (builtin DOOM modules vs. user packages).

With the version prior, only the contents of each file in `initFiles` was passed
to `initParser`, which made per-file behavior much harder to implement. By
reading inside the parser with the name as argument, this becomes easier.

I also added a direnv shell with rnix-lsp, you can safely cherry-pick that one
out if it isn't desired.

I thought this could help others, but if you think it won't, no worries, I'l try
and maintain my changes separately.

## Testing

I couldn't quite figure out how this is tested, what I ended up doing was
building the test folder's `flake.nix` and running the generated emacs
environment. That worked well, but it might not be sufficient to guarantee
accuracy.

### CODE REVIEW

**Comment by akirak** [comment]
Thank you for sending this PR. At first sight, the PR looked fine, but I have reverted that. I'm sorry for that.

Let me think more about this. I have understood your problem, but I don't like the idea of making `parseUsePackages` and `parseSetup` have an extra step of reading a file. It is less modular.

How about passing `initReader` as an option instead of `initParser`? That is, it would be better if the user could build his/her own reader by composing `builtins.readFile` and one of the parsers that take string as an argument. This way, `initFiles` actually doesn't have to be a list of files, so it will be more flexible.

Regarding the addition of `shell.nix`, I am not sure if I want to do that. I want to keep the workflow consistent across multiple projects as much as possible, so I want to try it out for some time before adopting it into the most crucial (?) project for me. I will consider using rnix-lsp. Thanks for the suggestion.

I tested this project by building [examples](https://github.com/emacs-twist/examples) with the input overridden (as below), but it takes more time. I add discovered corner cases to init.el in this repository, so it should work for most packages. The test suite needs improvement.

```sh
nix flake lock --update-input twist --override-input twist github:LuigiPiucco/twist.nix
```

**Comment by LuigiPiucco** [comment]
> Thank you for sending this PR. At first sight, the PR looked fine, but I have reverted that. I'm sorry for that.

No need to be sorry, discussing things is key to open-source, in my opinion.

> How about passing `initReader` as an option instead of `initParser`? That is, it would be better if the user could build his/her own reader by composing `builtins.readFile` and one of the parsers that take string as an argument. This way, `initFiles` actually doesn't have to be a list of files, so it will be more flexible.

I think it would work as well. Just to check if I understood correctly, is it something like this?

```nix
# In "pkgs/emacs/default.nix"
{
  # ...
  initParser ? lib.parseUsePackages {},
  initReader ? file: initParser (builtins.readFile file),
  # ...
}:
# ...
      (map (file: initReader file))
# Default parsers remain as they were.
```

If so, I can make such a change, I just need to know if you prefer that I add a new commit or rebase and force-push to my fork.

> Regarding the addition of `shell.nix`, I am not sure if I want to do that. I want to keep the workflow consistent across multiple projects as much as possible, so I want to try it out for some time before adopting it into the most crucial (?) project for me. I will consider using rnix-lsp. Thanks for the suggestion.

This was more for my own use, you can cherry-pick that one out (or, if you prefer, I can rebase without it). The case for having it defined per project is reproducibility: Anyone with direnv can have the proper environment, without having to install to the system globally each time.

**Comment by akirak** [comment]
> is it something like this?

Yes, I was thinking of a similar one.

> If so, I can make such a change, I just need to know if you prefer that I add a new commit or rebase and force-push to my fork.

Thanks. I don't mind either way, but force-push would be better.

> Anyone with direnv can have the proper environment, without having to install to the system globally each time.

Yes, I understand the point and use direnv. I use flake.nix to manage environments for my personal projects, but it adds extra inputs to flake.lock, which is not good for libraries. I have also recently discovered [devenv](https://github.com/cachix/devenv), so I haven't decided which one to use.

**Review by akirak — COMMENTED** [review]
I have made a trivial comment on the source. After you respond on it, I will merge this PR. Thank you for your contribution.

I was thinking of adding Doom to [examples](https://github.com/emacs-twist/examples). Hope to see your Nix configuration soon! It will make my project useful to others.

I tested this project by building the examples with the input overridden (as below). It takes more time to fetch and build sources for the full configurations, so this step is optional (or you can save time by running it with `--dry-run`). The test suite needs improvement.

```sh
nix build github:emacs-twist/examples#terlar.sandboxed  \
  --override-input twist github:LuigiPiucco/twist.nix \
  --impure --dry-run
```

For building, some packages require workarounds ([latest](https://git.sr.ht/~akirak/nix-config/tree/master/item/emacs/overrides.nix) and [slightly older on GitHub](https://github.com/akirak/nix-config/blob/migration/emacs/overrides.nix)). Some packages also have missing dependencies (which are upstream issues), which can be workaround like [this](https://github.com/akirak/nix-config/blob/migration/emacs/inputs.nix). I want to provide the workarounds as an independent repository, and it may entail refinement of the configuration API. I will create an issue for this.

File: pkgs/emacs/default.nix [hunk 1]
```diff
@@ -40,7 +41,7 @@ in
     archiveLockFile = lockDir + "/archive.lock";
 
     userConfig = lib.pipe self.initFiles [
-      (map (file: initParser (readFile file)))
+      (map (file: initReader file))
```
[the remark this answers is not in the document]
Comment (LuigiPiucco) [reply pkgs/emacs/default.nix:44]: Ah yes, good catch.

**Comment by LuigiPiucco** [comment]
> I was thinking of adding Doom to [examples](https://github.com/emacs-twist/examples). Hope to see your Nix configuration soon! It will make my project useful to others.

Sure! I am going to share it, and having an example would definitely be good. I just don't know how soon that'll be; we're at the end of semester period here, which is always very time-consuming. Once I get it working and minimally useful, I'll open-source it.

**Comment by akirak** [comment]
> I just don't know how soon that'll be; we're at the end of semester period here, which is always very time-consuming.

I don't mind that. I need to prepare user documentation for twist before it gets more public attention, but for some reasons I cannot work on it right now. 

Thank you for your contribution!
